app/src/main/java/com/mikhailovskii/weatherandroid/data/firebase/FirebaseModel.kt

```diff
@@ -0,0 +1,141 @@
+                val stickerPack = document.toObject(StickerPack::class.java)
+                stickerPackList.add(stickerPack)
+            }
+
+            callback?.onFirebaseDataLoaded(stickerPackList)
+        }.addOnFailureListener {
+            callback?.onFirebaseDataFailed()
+        }
+
+    }
+
+    fun saveUserLocation(user: User?) {
+        db.collection(USERS_COLLECTION).whereEqualTo(LOGIN_FIELD, user?.login).get()
+            .addOnSuccessListener { databaseResult ->
+
+                val map: Map<String, String> = hashMapOf(LOCATION to (user?.location ?: ""))
+
+                databaseResult.forEach { document ->
+                    db.collection(USERS_COLLECTION).document(document.id).update(map)
+                }
+
+            }
+    }
+
+    fun logInWithPassword(login: String, password: String, callback: FirebaseLoginCallback?) {
+        db.collection(USERS_COLLECTION).get().addOnSuccessListener { result ->
+            result.forEach { document ->
+                var user = document.toObject(User::class.java)
+
+                if (user.login == login && user.password == password) {
+                    user = User(login = login, password = password)
+
+                    if (user.location?.isNotBlank() == true) {
+                        callback?.onLoggedInWithFilledInfo()
+                    } else {
+                        callback?.onLoggedInWithEmptyLocation()
+                    }
+                }
+            }
+        }.addOnFailureListener {
+            callback?.onLoginFailed()
+        }
+    }
+
+    fun logInWithSocialNetwork(
+        social: String,
+        user: User,
+        callback: FirebaseLoginCallback?,
+        result: Result<com.twitter.sdk.android.core.models.User>?
+    ) {
+        db.collection(USERS_COLLECTION).get().addOnSuccessListener { databaseResult ->
+            var isUserPresent = false
+
+            databaseResult.forEach { document ->
+                val databaseUser = document.toObject(User::class.java)
+
+                val twitterCondition = databaseUser.login == result?.data?.name
+                        && databaseUser.twitterKey == result?.data?.idStr
+                        && databaseUser.icon == result?.data?.profileImageUrl
+
+                val facebookCondition = databaseUser.login == user.login
+                        && databaseUser.facebookKey == user.facebookKey
+                        && databaseUser.icon == user.icon
+
+                val googleCondition = databaseUser.login == user.login
+                        && databaseUser.googleKey == user.googleKey
+                        && databaseUser.icon == user.icon
+
+                if ((social == TWITTER && twitterCondition)
+                    || (social == FACEBOOK && facebookCondition)
+                    || (social == GOOGLE && googleCondition)
+                ) {
+                    isUserPresent = true
+                    return@forEach
+                }
```

Comment:
@SergeiMikhailovskii this is not what I meant saying "think how to unify logic from logInWithTwitter, logInWithFacebook and logInWithGoogle". You'd better have 3 different methods (as you used to have before update the PR) and a private one that handles the common stuff. As a lambda parameter you may pass into this method what should be done differently. Try to restore your previous implementation and write what I'm saying. It would look much more readable and clear

Reply:
@anton-novikau but as I see the only difference between all this social authorization is that I compare different fields of `User` entity. And it is done inside loop, which is common for all authorizations. That is why probably I don't understand how can I divide this method into 3

Reply:
You think too straightforward. How about this?
```kotlin
    fun logInWithTwitter(
        result: Result<com.twitter.sdk.android.core.models.User>?,
        user: User,
        callback: FirebaseLoginCallback?
    ) {
		logInWithSocialNetwork(user, callback) { databaseUser -> 
            return (databaseUser.login == result?.data?.name
                && databaseUser.twitterKey == result?.data?.idStr
                && databaseUser.icon == result?.data?.profileImageUrl
		}
    }
	
    fun logInWithGoogle(user: User, callback: FirebaseLoginCallback?) {
		logInWithSocialNetwork(user, callback) { databaseUser ->
			return databaseUser.login == user.login
                    && databaseUser.googleKey == user.googleKey
                    && databaseUser.icon == user.icon
		}
    }
	
    private fun logInWithSocialNetwork(user: User, callback: FirebaseLoginCallback?, checkUser: (user: User) -> Boolean) {
        db.collection(USERS_COLLECTION).get().addOnSuccessListener { databaseResult ->
            var isUserPresents = false

            databaseResult.forEach { document ->
		if (checkUser(document.toObject(User::class.java))) {
			isUserPresent = true
			return@forEach
		}
            }

            if (!isUserPresents) {
                db.collection(USERS_COLLECTION).document().set(user)
                    .addOnSuccessListener {
                        Timber.d("Social Network info saved")
                    }.addOnFailureListener { e ->
                        Timber.e("Error Social Network save: $e")
                    }
            }

            callback?.onLoggedInWithFilledInfo()
        }.addOnFailureListener {
            callback?.onLoginFailed()
        }	
    }
```